crates/vfio-ioctls/src/vfio_device.rs

```diff
@@ -426,6 +278,108 @@ impl VfioContainer {
     }
 }
 
+#[cfg(all(feature = "kvm", not(test)))]
```

Comment:
Can you add some comments about the design choices you are making here? For example, that for testing purposes you are having separate implementations for certain functions that otherwise are platform dependent? These kind of things tend to get lost in the commit messages, and people won't know when looking at the code 2 years from now why things are the way they are.

Reply:
Add some comments:
```
#[cfg(all(feature = "kvm", not(test)))]
// Methods to support the KVM hypervisor.
// Note: a special stub implementation is used for VFIO unit tests, so following code won't covered
// by unit tests, be careful when review changes.
```
#[cfg(any(test, all(not(feature = "mshv"), not(feature = "kvm"))))]
// Methods to support user mode driver, which has no associated hypervisors.
// This implementation also acts a stub for VFIO unit tests to avoid dependency on platform
// hardware configuration.
```

```

Reply:
I still think it's a better idea to make them unsafe, because we are not doing any checks there for the return value of the unsafe code.

Reply:
That will cause a big messy:(
It's easy to mark the ioctl based implementation as unsafe, but that will cause trouble to the mock version.
For example, if we mark 
```
    pub fn check_api_version(file: &File) -> i32 {
        unsafe { ioctl(file, VFIO_GET_API_VERSION()) }
    }
```
as unsafe, then we need also to mark
```
    pub fn check_api_version(_file: &File) -> i32 {
        VFIO_API_VERSION as i32
    }
```
as unsafe too.